apis/apiextensions/v1/composition_environment.go

```diff
@@ -67,12 +92,49 @@ type EnvironmentSource struct {
 	Selector *EnvironmentSourceSelector `json:"selector,omitempty"`
 }
 
+// Validate the EnvironmentSource.
+func (e *EnvironmentSource) Validate() *field.Error {
+	switch e.Type {
+	case EnvironmentSourceTypeReference:
+		if e.Ref == nil {
+			return field.Required(field.NewPath("ref"), "ref is required")
+		}
+		if err := e.Ref.Validate(); err != nil {
+			return errors.WrapFieldError(err, field.NewPath("ref"))
+		}
+
+	case EnvironmentSourceTypeSelector:
+		if e.Selector == nil {
+			return field.Required(field.NewPath("selector"), "selector is required")
+		}
+		if len(e.Selector.MatchLabels) == 0 {
+			return field.Required(field.NewPath("selector", "matchLabels"), "selector must have at least one match label")
+		}
+		for i, m := range e.Selector.MatchLabels {
+			if err := m.Validate(); err != nil {
+				return errors.WrapFieldError(err, field.NewPath("selector", "matchLabels").Index(i))
+			}
+		}
+	default:
+		return field.Invalid(field.NewPath("type"), e.Type, "invalid type")
+	}
+	return nil
+}
+
 // An EnvironmentSourceReference references an EnvironmentConfig by it's name.
 type EnvironmentSourceReference struct {
 	// The name of the object.
 	Name string `json:"name"`
 }
 
+// Validate the EnvironmentSourceReference.
+func (e *EnvironmentSourceReference) Validate() *field.Error {
+	if e.Name == "" {
+		return field.Required(field.NewPath("name"), "name is required")
```

Comment:
IMHO, we should define this constraint on CRD schema, so that k8s api server can verify it without a need to write any code on our side?

Reply:
Yes, this one in particular could be enforced by setting min length to 1, which I did in the beginning, but then it would only return that and any other error according to the schema first, and only after return any other error from the validation logic, so I preferred having them checked at a single level. Most of the other ones would require CEL rules given that are mostly depending on the "type", which we could add, but I prefer having them in code until we don't put in place E2Es able to validate them.

Reply:
This is something we discussed [early](https://github.com/crossplane/crossplane/pull/3921#discussion_r1150817325). I remember we removed them first but added them back for some reason, probably to be consistent with the validations that the API server cannot do.